File: src/main/java/com/ject/studytrip/auth/infra/provider/KakaoOauthProvider.java
package com.ject.studytrip.auth.infra.provider;

import static org.springframework.util.StringUtils.hasText;

import com.ject.studytrip.auth.domain.error.AuthErrorCode;
import com.ject.studytrip.auth.infra.client.KakaoOauthClient;
import com.ject.studytrip.auth.infra.dto.KakaoTokenResponse;
import com.ject.studytrip.auth.infra.dto.KakaoUserInfoResponse;
import com.ject.studytrip.global.config.properties.KakaoOauthProperties;
import com.ject.studytrip.global.exception.CustomException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.BodyInserters;

@Component
@RequiredArgsConstructor
public class KakaoOauthProvider {
    private final KakaoOauthClient kakaoOauthClient;
    private final KakaoOauthProperties kakaoOauthProperties;

    public KakaoTokenResponse getKakaoTokens(String code) {
        validateKakaoAuthorizationCode(code);
        return kakaoOauthClient
                .fetchKakaoTokens(kakaoOauthProperties.tokenUri(), createFormData(code))
                .block();
    }

    public KakaoUserInfoResponse getKakaoUserInfo(String accessToken) {
        validateKakaoToken(accessToken);
        return kakaoOauthClient
                .fetchKakaoUserInfo(kakaoOauthProperties.userInfoUri(), accessToken)
                .block();
    }

    private BodyInserters.FormInserter<String> createFormData(String code) {
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("grant_type", "authorization_code");
        formData.add("client_id", kakaoOauthProperties.clientId());
        formData.add("client_secret", kakaoOauthProperties.clientSecret());
        formData.add("redirect_uri", kakaoOauthProperties.redirectUri());
        formData.add("code", code);
        return BodyInserters.fromFormData(formData);
    }

    private void validateKakaoAuthorizationCode(String code) {
        if (!hasText(code)) {
            throw new CustomException(AuthErrorCode.INVALID_KAKAO_AUTHORIZATION_CODE);
        }
    }

    private void validateKakaoToken(String accessToken) {
        if (!hasText(accessToken)) {
            throw new CustomException(AuthErrorCode.INVALID_KAKAO_TOKEN);
        }
    }
}

File: src/main/kotlin/com/ject/studytrip/auth/infra/provider/KakaoOauthProvider.kt
package com.ject.studytrip.auth.infra.provider

import com.ject.studytrip.auth.domain.error.AuthErrorCode
import com.ject.studytrip.auth.infra.client.KakaoOauthClient
import com.ject.studytrip.auth.infra.dto.KakaoTokenResponse
import com.ject.studytrip.auth.infra.dto.KakaoUserInfoResponse
import com.ject.studytrip.global.config.properties.KakaoOauthProperties
import com.ject.studytrip.global.exception.CustomException
import org.springframework.stereotype.Component
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.StringUtils.hasText
import org.springframework.web.reactive.function.BodyInserters

@Component
class KakaoOauthProvider(
    private val kakaoOauthClient: KakaoOauthClient,
    private val kakaoOauthProperties: KakaoOauthProperties,
) {
    fun getKakaoTokens(
        code: String,
        origin: String,
    ): KakaoTokenResponse {
        validateKakaoAuthorizationCode(code)

        return kakaoOauthClient
            .fetchKakaoTokens(kakaoOauthProperties.tokenUri, createFormData(code, origin))
            .block()
            ?: throw CustomException(AuthErrorCode.INVALID_KAKAO_AUTHORIZATION_CODE)
    }

    fun getKakaoUserInfo(accessToken: String): KakaoUserInfoResponse {
        validateKakaoToken(accessToken)

        return kakaoOauthClient
            .fetchKakaoUserInfo(kakaoOauthProperties.userInfoUri, accessToken)
            .block()
            ?: throw CustomException(AuthErrorCode.INVALID_KAKAO_TOKEN)
    }

    private fun createFormData(
        code: String,
        origin: String,
    ) = BodyInserters.fromFormData(
        LinkedMultiValueMap<String, String>().apply {
            add("grant_type", "authorization_code")
            add("client_id", kakaoOauthProperties.clientId)
            add("client_secret", kakaoOauthProperties.clientSecret)
            add("redirect_uri", origin + kakaoOauthProperties.redirectUri)
            add("code", code)
        },
    )

    private fun validateKakaoAuthorizationCode(code: String) {
        if (!hasText(code)) {
            throw CustomException(AuthErrorCode.INVALID_KAKAO_AUTHORIZATION_CODE)
        }
    }

    private fun validateKakaoToken(accessToken: String) {
        if (!hasText(accessToken)) {
            throw CustomException(AuthErrorCode.INVALID_KAKAO_TOKEN)
        }
    }
}
